# 2. teden: Numpy - Domača naloga

Uvozimo modul numpy:

In [2]:
import numpy as np

### 1.Create a null vector of size 10

In [3]:
x = np.zeros(10)
print(x)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [5]:
x = np.zeros(10)
x[4] = 1
print(x)

[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [18]:
x = 10 + np.arange(40)
print (x)

[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]


Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [6]:
z = np.arange(50)
z = z[::-1]
print(z)

[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26
 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2
  1  0]


Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [26]:
x = np.arange(9).reshape(3,3)
print(x)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [9]:
a = [1,2,0,0,4,0]
np.nonzero(a)


(array([0, 1, 4]),)

Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [10]:
x = np.eye(3)
print(x)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [11]:
x = np.random.random((3,3,3))
print(x)

[[[0.28265862 0.63499512 0.70110279]
  [0.03825501 0.79524396 0.08016499]
  [0.41406279 0.51311325 0.6561952 ]]

 [[0.25718907 0.57196974 0.550878  ]
  [0.70034331 0.03893617 0.91059611]
  [0.9946557  0.01403936 0.80835681]]

 [[0.65870809 0.53137841 0.81823834]
  [0.07975337 0.86957354 0.39024462]
  [0.89331443 0.08167743 0.98386162]]]


Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [12]:
x = np.random.random((10,10))
xmin, xmax = x.min(), x.max()

a = print (xmin)
b = print (xmax)


0.0016930649017564692
0.9991861896002837


Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [13]:
x = np.ones((10,10))
x[1:-1,1:-1] = 0
print(x)

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

In [77]:
0 * np.nan

nan

In [4]:
np.nan == np.nan

False

In [23]:
np.inf > np.nan


False

In [5]:
np.nan - np.nan

nan

In [81]:
0.3 == 3 * 0.1 

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [7]:
x = np.ones((3,3))
x = np.zeros((8,8),dtype=int)
x[1::2,::2] = 1
x[::2,1::2] = 1
print(x)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [14]:
x = [0,1,2,3,4,5,6,7,8,9,10]
x = np.arange(11)
x[(3 < x) & (x < 8)] *= -1
print(x)

[ 0  1  2  3 -4 -5 -6 -7  8  9 10]


Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [15]:
x = np.linspace(0,1,12,endpoint=True)[1:-1]
print(x)

[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455
 0.63636364 0.72727273 0.81818182 0.90909091]


Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [28]:
x = np.random.random(10)
print("*original:")
print(x)
print ("---------------------------------------------------------------------------------------------------------------")
x.sort()
print("**modified:")
print(x)

*original:
[0.01563476 0.03174178 0.4225454  0.12323724 0.12582823 0.77902121
 0.54350116 0.12554938 0.67904864 0.92188321]
---------------------------------------------------------------------------------------------------------------
**modified:
[0.01563476 0.03174178 0.12323724 0.12554938 0.12582823 0.4225454
 0.54350116 0.67904864 0.77902121 0.92188321]


Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [15]:
A = np.random.randint(0,2,5)

In [16]:
B = np.random.randint(0,2,5)

In [17]:
A = np.random.randint(0,2,5)
print("array A:")
print(A)
B = np.random.randint(0,2,5)
print("array B:")
print(B)
array_equal = np.allclose(A, B)
print(array_equal)

array A:
[0 1 0 0 0]
array B:
[0 1 1 0 1]
False


### 17.Make an array immutable (read-only)

In [40]:
Z = np.zeros(10)

In [7]:
Z = np.zeros(10)
Z.flags.writeable = False
Z[2] = 5

ValueError: assignment destination is read-only

### 18.Create random vector of size 10 and replace the maximum value by 0

In [18]:
x = np.random.random(10)
print("*original:")
print(x)
print ("---------------------------------------------------------------------------------------------------------------")
x[x.argmax()] = 0
print("**modified:")
print(x)


*original:
[0.23123077 0.4657852  0.59652588 0.67334041 0.97841592 0.54116011
 0.61927422 0.54908196 0.08308404 0.21367344]
---------------------------------------------------------------------------------------------------------------
**modified:
[0.23123077 0.4657852  0.59652588 0.67334041 0.         0.54116011
 0.61927422 0.54908196 0.08308404 0.21367344]


### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [18]:
Z = np.arange(100)

In [20]:
v = np.random.uniform(0,100)

In [20]:
x = np.arange(100)
print("*original:")
print(x)
print ("---------------------------------------------------------------------------------------------------------------")
a = np.random.uniform(0,100)
print("**closest value to v:")
print(a)
index = (np.abs(x-a)).argmin()
print(x[index])

*original:
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]
---------------------------------------------------------------------------------------------------------------
**closest value to v:
67.3209190683422
67


### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [19]:
Z = np.arange(10, dtype=np.int32)
import numpy as np
Z = np.arange(10, dtype='int32')
X = Z.view('float32')
X[:] = Z

print(X)



[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]


Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

In [21]:
Z = np.random.randint(0,10,50)
print(Z)


[3 7 4 3 7 0 6 3 4 3 8 0 5 9 4 2 7 5 1 0 3 0 3 9 3 7 7 5 0 4 9 3 1 9 6 9 9
 4 3 1 2 7 7 7 4 5 2 4 3 0]
